polynomial.hpp

```diff
@@ -0,0 +1,220 @@
+                        break;
+                    }
+                }
+                if(duplicates == 0){
+                    factors.push_back(factors_of_P[i] / factors_of_Q[j]);
+                }
+                duplicates = 0;
+            }
+        }
+        size_t results_positive = 0;
+        size_t results_negative = 0;
+        for(std::size_t i = 0; i < factors.size; i++){
+            for(std::size_t j = 0; j < p.degree; j++){
+                results_positive += p.coefficients[j] * factors[i];
+                results_negative += p.coefficients[j] * -factors[i];
+            }
+            if(results_positive == 0){
+                return factors[i];
+            }else if(results_negative == 0){
+                return -factors[i];
+            }
+            results_positive = 0;
+            results_negative = 0;
+        }
+        throw std::invalid_argument("Polynomial has no rational roots");
+    }
+    template <typename T>
+    inline auto gcd(polynomial<T> p, polynomial<T> q) -> std::pair<polynomial<T>, std::vector<T>>
+    {
+	std::pair<polynomial<T>, std::vector<T>> result;
+	polynomial<T> x_prev{1}, y_prev{0}, x{0}, y{1};
+	std::vector<T> results;
+    while (q.degree >= 0)
+    {
+        auto quotient = divide(p, q).first;
+        p -= quotient * q;
+        auto temp = std::move(x);
+        x = x_prev - quotient * x;
+        x_prev = std::move(temp);
+        temp = std::move(y);
+        y = y_prev - quotient * y;
+        y_prev = std::move(temp);
+        results.push_back(x_prev);
+        results.push_back(y_prev);
+    }
+    result.first = x_prev; 
+    result.second = std::move(results); 
+    return result;
+    }
+}
+void test() {
+    polynomial::polynomial<double> p1{{1, 2, 3}};
+    polynomial::polynomial<double> p2{{1, 1}};
+    polynomial::polynomial<double> result_addition = p1 + p2;
+    std::cout << "Addition result: ";
+    for (auto coeff : result_addition.coefficients) {
+        std::cout << coeff << " ";
+    }
+    std::cout << std::endl;
+    polynomial::polynomial<double> result_subtraction = p1 - p2;
+    std::cout << "Subtraction result: ";
+    for (auto coeff : result_subtraction.coefficients) {
+        std::cout << coeff << " ";
+    }
+    std::cout << std::endl;
+    polynomial::polynomial<double> result_multiplication = p1 * p2;
+    std::cout << "Multiplication result: ";
+    for (auto coeff : result_multiplication.coefficients) {
+        std::cout << coeff << " ";
+    }
+    std::cout << std::endl;
+    polynomial::polynomial<double> result_division = p1 / p2;
+    std::cout << "Division result: ";
+    for (auto coeff : result_division.coefficients) {
+        std::cout << coeff << " ";
+    }
+    std::cout << std::endl;
+    polynomial::polynomial<double> result_modulus = p1 % p2;
+    std::cout << "Modulus result: ";
+    for (auto coeff : result_modulus.coefficients) {
+        std::cout << coeff << " ";
+    }
+    std::cout << std::endl;
+}
```

Comment:
you have much more fucntions to test - value of polynomial evaluated on a given arguement and gcd